docs/config.md
# Configuration options

You can either use a `config.yml` file in the root directory of vikunja or set all config option with 
environment variables. If you have both, the value set in the config file is used.

Variables are nested in the `config.yml`, these nested variables become `VIKUNJA_FIRST_CHILD` when configuring via
environment variables. So setting

```bash
export VIKUNJA_FIRST_CHILD=true
```

is the same as defining it in a `config.yml` like so:

```yaml
first:
    child: true
```

## Config file locations

Vikunja will search on various places for a config file:

* Next to the location of the binary
* In the `service.rootpath` location set in a config (remember you can set config arguments via environment variables)
* In `/etc/vikunja`
* In `~/.config/vikunja`

# Default configuration with explanations

This is the same as the `config.yml.sample` file you'll find in the root of vikunja.

```yaml
service:
  # This token is used to verify issued JWT tokens.
  # Default is a random token which will be generated at each startup of vikunja.
  # (This means all already issued tokens will be invalid once you restart vikunja)
  JWTSecret: "cei6gaezoosah2bao3ieZohkae5aicah"
  # The interface on which to run the webserver
  interface: ":3456"
  # The URL of the frontend, used to send password reset emails.
  frontendurl: ""
  # The base path on the file system where the binary and assets are.
  # Vikunja will also look in this path for a config file, so you could provide only this variable to point to a folder
  # with a config file which will then be used.
  rootpath: <the path of the executable>
  # The number of items which gets returned per page
  pagecount: 50
  # If set to true, enables a /metrics endpoint for prometheus to collect metrics about the system
  # You'll need to use redis for this in order to enable common metrics over multiple nodes
  enablemetrics: false

database:
  # Database type to use. Supported types are mysql and sqlite.
  type: "sqlite"
  # Database user which is used to connect to the database.
  user: "vikunja"
  # Databse password
  password: ""
  # Databse host
  host: "localhost"
  # Databse to use
  database: "vikunja"
  # When using sqlite, this is the path where to store the data
  Path: "./vikunja.db"
  # Sets the max open connections to the database. Only used when using mysql.
  openconnections: 100

cache:
  # If cache is enabled or not
  enabled: false
  # Cache type. Possible values are memory or redis, you'll need to enable redis below when using redis
  type: memory
  # When using memory this defines the maximum size an element can take
  maxelementsize: 1000

redis:
  # Whether to enable redis or not
  enabled: false
  # The host of the redis server including its port.
  redishost: 'localhost:6379'
  # The password used to authenicate against the redis server
  redispassword: ''
  # 0 means default database
  db: 0

mailer:
  # Whether to enable the mailer or not. If it is disabled, all users are enabled right away and password reset is not possible.
  enabled: false
  # SMTP Host
  host: ""
  # SMTP Host port
  port: 587
  # SMTP username
  username: "user"
  # SMTP password
  password: ""
  # Wether to skip verification of the tls certificate on the server
  skiptlsverify: false
  # The default from address when sending emails
  fromemail: "mail@vikunja"
  # The length of the mail queue.
  queuelength: 100
  # The timeout in seconds after which the current open connection to the mailserver will be closed.
  queuetimeout: 30

log:
  # A folder where all the logfiles should go.
  path: <rootpath>logs
  # Whether to show any logging at all or none
  enabled: true
  # Where the error log should go. Possible values are stdout, stderr, file or off to disable error logging.
  errors: "stdout"
  # Where the normal log should go. Possible values are stdout, stderr, file or off to disable standard logging.
  standard: "stdout"
  # Whether or not to log database queries. Useful for debugging. Possible values are stdout, stderr, file or off to disable database logging.
  database: "off"
  # Whether to log http requests or not. Possible values are stdout, stderr, file or off to disable http logging.
  http: "stdout"
  # Echo has its own logging which usually is unnessecary, which is why it is disabled by default. Possible values are stdout, stderr, file or off to disable standard logging.
  echo: "off"
```


pkg/config/config.go
//  Vikunja is a todo-list application to facilitate your life.
//  Copyright 2018 Vikunja and contributors. All rights reserved.
//
//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <https://www.gnu.org/licenses/>.

package config

import (
	"code.vikunja.io/api/pkg/log"
	"crypto/rand"
	"fmt"
	"github.com/spf13/viper"
	"os"
	"path/filepath"
	"strings"
)

// InitConfig initializes the config, sets defaults etc.
func init() {

	// Set defaults
	// Service config
	random, err := random(32)
	if err != nil {
		log.Log.Fatal(err.Error())
	}

	// Service
	viper.SetDefault("service.JWTSecret", random)
	viper.SetDefault("service.interface", ":3456")
	viper.SetDefault("service.frontendurl", "")

	ex, err := os.Executable()
	if err != nil {
		panic(err)
	}
	exPath := filepath.Dir(ex)
	viper.SetDefault("service.rootpath", exPath)
	viper.SetDefault("service.pagecount", 50)
	viper.SetDefault("service.enablemetrics", false)
	// Database
	viper.SetDefault("database.type", "sqlite")
	viper.SetDefault("database.host", "localhost")
	viper.SetDefault("database.user", "vikunja")
	viper.SetDefault("database.password", "")
	viper.SetDefault("database.database", "vikunja")
	viper.SetDefault("database.path", "./vikunja.db")
	viper.SetDefault("database.openconnections", 100)
	// Cacher
	viper.SetDefault("cache.enabled", false)
	viper.SetDefault("cache.type", "memory")
	viper.SetDefault("cache.maxelementsize", 1000)
	// Mailer
	viper.SetDefault("mailer.enabled", false)
	viper.SetDefault("mailer.host", "")
	viper.SetDefault("mailer.port", "587")
	viper.SetDefault("mailer.user", "user")
	viper.SetDefault("mailer.password", "")
	viper.SetDefault("mailer.skiptlsverify", false)
	viper.SetDefault("mailer.fromemail", "mail@vikunja")
	viper.SetDefault("mailer.queuelength", 100)
	viper.SetDefault("mailer.queuetimeout", 30)
	// Redis
	viper.SetDefault("redis.enabled", false)
	viper.SetDefault("redis.host", "localhost:6379")
	viper.SetDefault("redis.password", "")
	viper.SetDefault("redis.db", 0)
	// Logger
	viper.SetDefault("log.enabled", true)
	viper.SetDefault("log.errors", "stdout")
	viper.SetDefault("log.standard", "stdout")
	viper.SetDefault("log.database", "off")
	viper.SetDefault("log.http", "stdout")
	viper.SetDefault("log.echo", "off")
	viper.SetDefault("log.path", viper.GetString("service.rootpath")+"/logs")

	// Init checking for environment variables
	viper.SetEnvPrefix("vikunja")
	viper.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
	viper.AutomaticEnv()

	// Load the config file
	viper.AddConfigPath(viper.GetString("service.rootpath"))
	viper.AddConfigPath("/etc/vikunja/")
	viper.AddConfigPath("~/.config/vikunja")
	viper.AddConfigPath(".")
	viper.SetConfigName("config")
	err = viper.ReadInConfig()
	if err != nil {
		log.Log.Info(err)
		log.Log.Info("Using defaults.")
	}
}

func random(length int) (string, error) {
	b := make([]byte, length)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}

	return fmt.Sprintf("%X", b), nil
}
